<!DOCTYPE html>
<html>

  <head>
    <meta charset='utf-8'>
    <meta http-equiv="X-UA-Compatible" content="chrome=1">
    <meta name="description" content="Gotcloud : Genomes on the Cloud, Mapping &amp; Variant Calling Pipelines">

    <link rel="stylesheet" type="text/css" media="screen" href="stylesheets/stylesheet.css">

    <title>Gotcloud</title>
  </head>

  <body>

    <!-- HEADER -->
    <div id="header_wrap" class="outer">
        <header class="inner">
          <a id="forkme_banner" href="https://github.com/statgen/gotcloud">View on GitHub</a>

          <h1 id="project_title">Gotcloud</h1>
          <h2 id="project_tagline">Genomes on the Cloud, Mapping &amp; Variant Calling Pipelines</h2>

            <section id="downloads">
              <a class="zip_download_link" href="https://github.com/statgen/gotcloud/zipball/master">Download this project as a .zip file</a>
              <a class="tar_download_link" href="https://github.com/statgen/gotcloud/tarball/master">Download this project as a tar.gz file</a>
            </section>
        </header>
    </div>

    <!-- MAIN CONTENT -->
    <div id="main_content_wrap" class="outer">
      <section id="main_content" class="inner">
        <h3>
<a name="welcome-to-gotcloud" class="anchor" href="#welcome-to-gotcloud"><span class="octicon octicon-link"></span></a>Welcome to GotCloud.</h3>

<p>Genomes on the Cloud (GotCloud) is our set of sequence analysis software pipelines that we developed to handle the increasing volume of next generation sequencing and genotyping data.</p>

<ul>
<li>Integrative - Alignment, QC, Variant Calling, Phasing</li>
<li>Seamless - Requires only simple configuration files

<ul>
<li>Fully automated and parallelized</li>
</ul>
</li>
<li>Robust against unexpected failures &amp; stops</li>
<li>Scalable from a few to many thousands of genomes

<ul>
<li>Deep Genomes, Shallow Genomes, Targeted Exomes</li>
</ul>
</li>
</ul>

<p>Also included within GotCloud</p>

<ul>
<li>Set of many useful software tools</li>
<li>Software C++ library for developing additional sequence analysis tools</li>
</ul>

<p>See (<a href="http://gotcloud.org">http://gotcloud.org</a>) for full GotCloud documentation.</p>

<h3>
<a name="obtain-gotcloud" class="anchor" href="#obtain-gotcloud"><span class="octicon octicon-link"></span></a>Obtain GotCloud</h3>

<p>First download GotCloud either using git or by using one of the download buttons to the left. 
See (<a href="http://genome.sph.umich.edu/wiki/GotCloud:_Source_Releases">http://genome.sph.umich.edu/wiki/GotCloud:_Source_Releases</a>) for more information</p>

<h3>
<a name="build-gotcloud" class="anchor" href="#build-gotcloud"><span class="octicon octicon-link"></span></a>Build GotCloud</h3>

<p>Before running GotCloud, you need to compile the source:</p>

<pre><code>$ cd your_downloaded_gotcloud_dir/src; make; cd ..
</code></pre>

<h3>
<a name="test-gotcloud" class="anchor" href="#test-gotcloud"><span class="octicon octicon-link"></span></a>Test GotCloud</h3>

<p>To test GotCloud align, run:</p>

<pre><code>$ ./gotcloud align --test ~/testalign
</code></pre>

<p>This will create/clear the output directory ~/testalign (alternatively, you can choose a different directory).
Test results and a log file are put into that directory.
Results are self-checked, and if any errors should occur, GotCloud will print an error message.</p>

<p>To test  GotCloud snpcall, run:</p>

<pre><code>$ ./gotcloud snpcall --test ~/testsnp
</code></pre>

<p>This will create/clear the output directory ~/testsnp (alternatively, you can choose a different directory).
Test results and a log file are put into that directory.
Results are self-checked, and if any errors should occur, GotCloud will print an error message.</p>

<p>To test  GotCloud ldrefine, run:</p>

<pre><code>$ ./gotcloud ldrefine --test ~/testsnp
</code></pre>

<p>The ldrefine test can only be run after the snpcall test, and must use the same directory used for the snpcall test.
Test results and a log file are put into that directory.
Results are self-checked, and if any errors should occur, GotCloud will print an error message.</p>

<p>To test  GotCloud indel, run:</p>

<pre><code>$ ./gotcloud indel --test ~/testindel
</code></pre>

<p>This will create/clear the output directory ~/testindel (alternatively, you can choose a different directory).
Test results and a log file are put into that directory.
Results are self-checked, and if any errors should occur, GotCloud will print an error message.</p>

<h3>
<a name="general-help-for-variant-calling" class="anchor" href="#general-help-for-variant-calling"><span class="octicon octicon-link"></span></a>General Help for Variant Calling</h3>

<p>Variant Calling requires three types of input files:
1. a set of BAM files</p>

<ul>
<li>For high quality SNP calls BAM files should already be:

<ul>
<li>duplicate-marked </li>
<li>base-quality recalibrated</li>
<li>BAM list file</li>
</ul>
</li>
<li>Each line contains at least 2-3 space-separated columns representing a single individual:
'''
  [SAMPLE_ID] [COMMA SEPARATED POPULATION LABELS] [BAM_FILE1] [BAM_FILE2] ...
'''

<ul>
<li>1 or more BAMs are allowed per individual</li>
<li>Population column is optional and can be skipped</li>
<li>configuration file</li>
</ul>
</li>
<li>Contains run-time options &amp; command line arguments</li>
<li>A default configuration is provided</li>
<li>User must specify (KEY = VALUE pairs):

<ul>
<li>BAM_LIST    # the path/name of the bam list file</li>
<li>OUT_DIR</li>
</ul>
</li>
<li>Optional specifications...

<ul>
<li>CHRS = #space separated list of chromosomes - defaults to 1-22 &amp; X</li>
</ul>
</li>
<li>Refer to the default configuration &amp; the wiki page for more information on other settings.

<ul>
<li><a href="http://genome.sph.umich.edu/wiki/GotCloud:_Variant_Calling_Pipeline#Configuration_File">http://genome.sph.umich.edu/wiki/GotCloud:_Variant_Calling_Pipeline#Configuration_File</a></li>
</ul>
</li>
</ul>

<p>Optional input files:</p>

<ul>
<li>Pedigree files (PED format) (to specify gender information in chrX calling)</li>
<li>Target information (UCSC's BED format) in targeted or whole exome capture sequencing</li>
</ul>

<p>Once these files are configured, to run snp calling and process the data:</p>

<pre><code>   {path}gotcloud snpcall --conf {conf_file} \
      --outdir {output_directory} --numjobs {# of threads to use for processing}
</code></pre>

<h3>
<a name="authors-and-contributors" class="anchor" href="#authors-and-contributors"><span class="octicon octicon-link"></span></a>Authors and Contributors</h3>

<p>GotCloud has been developed and maintained by within the Center of Statistical Genetics at the University of Michigan.  The following people have worked on and contributed to GotCloud:</p>

<ul>
<li>Goncalo Abecasis</li>
<li>Hyun Min Kang</li>
<li>Goo Jun</li>
<li>Mary Kate Wing</li>
<li>Terry Gliedt</li>
<li>Chris Scheller</li>
<li>Adrian Tan</li>
<li>Alan Kwong</li>
<li>Tom Blackwell</li>
</ul>

<p>The Group's github account is <a href="https://github.com/statgen" class="user-mention">@statgen</a>.</p>

<h3>
<a name="support-or-contact" class="anchor" href="#support-or-contact"><span class="octicon octicon-link"></span></a>Support or Contact</h3>

<p>Having trouble with GotCloud?  </p>

<ul>
<li>Check out the documentation at gotcloud.org </li>
<li>See <a href="http://genome.sph.umich.edu/wiki/GotCloud:_FAQs">GotCloud: FAQs</a>
</li>
<li>See <a href="http://groups.google.com/group/GotCloud">GotCloud Google Group</a> to view past questions/comments or to ask your question

<ul>
<li>Join the GotCloud Google Group to post a question or to be notified of new releases</li>
</ul>
</li>
<li>E-mail Mary Kate Wing (<a href="mailto:mktrost@umich.edu">mktrost@umich.edu</a>).</li>
</ul>
      </section>
    </div>

    <!-- FOOTER  -->
    <div id="footer_wrap" class="outer">
      <footer class="inner">
        <p class="copyright">Gotcloud maintained by <a href="https://github.com/statgen">statgen</a></p>
        <p>Published with <a href="http://pages.github.com">GitHub Pages</a></p>
      </footer>
    </div>

    

  </body>
</html>
